### ISSUE TITLE
Help with BIOMOD_xxx - [every time I run the models the outputs change]

### ISSUE THREAD

**Issue by zali37** [biomodhub/biomod2#643]
Hello biomod team

**Context and question**
I have a question about the stability of the models' outputs. I think if I am doing things well, when I rerun the models, I should obtain the same results, but it is not the case, and I do not know what I am doing wrong. I am working with two species. One of them is always in retraction, and the other is always in expansion, but the current and future range size of each always changes. 

**Code used**
darwinii_format <- BIOMOD_FormatingData(
  resp.name = "darwinii",
  resp.var = darwinii$presence,
  resp.xy = darwinii[, c("lon", "lat")],
  expl.var = stk_current,
  eval.resp.var = darwinii_eval$presence,
  eval.expl.var = stk_current,
  eval.resp.xy = darwinii_eval[, c("lon", "lat")],
  filter.raster = T,
  na.rm = T
)
-=-=-=-=-=-=-=-=-=-=-=-=-=-=-= BIOMOD.formated.data -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=

dir.name =  .

sp.name =  darwinii

	 24 presences,  39 true absences and  0 undefined points in dataset

	 7 explanatory variables

   clip_bio3       clip_bio6        clip_bio8        clip_bio9      
 Min.   :37.62   Min.   :-7.068   Min.   :-2.930   Min.   : 0.2247  
 1st Qu.:45.19   1st Qu.:-3.236   1st Qu.: 2.711   1st Qu.: 4.7909  
 Median :46.98   Median :-1.996   Median : 5.165   Median : 7.3880  
 Mean   :45.86   Mean   :-1.966   Mean   : 5.138   Mean   : 8.1280  
 3rd Qu.:47.65   3rd Qu.:-1.028   3rd Qu.: 7.488   3rd Qu.:12.1411  
 Max.   :49.85   Max.   : 1.752   Max.   :11.746   Max.   :16.3493  
   clip_bio14       clip_bio15     pedregal_ULTIMO3
 Min.   :  4.00   Min.   : 8.682   Min.   :0.0000  
 1st Qu.: 12.00   1st Qu.:12.362   1st Qu.:0.0000  
 Median : 50.00   Median :19.799   Median :0.0000  
 Mean   : 73.65   Mean   :21.445   Mean   :0.1429  
 3rd Qu.:135.50   3rd Qu.:26.382   3rd Qu.:0.0000  
 Max.   :194.00   Max.   :56.824   Max.   :1.0000  

Evaluation data :

	 9 presences,  18 true absences and  0 undefined points in dataset

   clip_bio3       clip_bio6         clip_bio8        clip_bio9     
 Min.   :36.99   Min.   :-4.5600   Min.   : 0.974   Min.   : 0.676  
 1st Qu.:42.70   1st Qu.:-2.7380   1st Qu.: 2.636   1st Qu.: 6.368  
 Median :46.31   Median :-1.4000   Median : 3.945   Median :10.215  
 Mean   :44.85   Mean   :-1.5617   Mean   : 5.006   Mean   : 9.335  
 3rd Qu.:47.15   3rd Qu.:-0.4875   3rd Qu.: 7.218   3rd Qu.:12.129  
 Max.   :49.60   Max.   : 0.8333   Max.   :13.856   Max.   :15.613  
   clip_bio14      clip_bio15     pedregal_ULTIMO3
 Min.   :  5.0   Min.   : 9.706   Min.   :0.0000  
 1st Qu.:  8.5   1st Qu.:21.905   1st Qu.:0.0000  
 Median : 15.0   Median :27.083   Median :0.0000  
 Mean   : 60.0   Mean   :28.810   Mean   :0.1852  
 3rd Qu.:119.5   3rd Qu.:36.245   3rd Qu.:0.0000  
 Max.   :183.0   Max.   :52.647   Max.   :1.0000  

-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
# Modeling options ########
cv.k <- bm_CrossValidation(bm.format = darwinii_format,
                           strategy = "kfold",
                           k = 5,
                           nb.rep = 10)

opt.d <- bm_ModelingOptions(data.type = 'binary',
                            models = c("RF","ANN","GAM.gam.gam","SRE"),
                            strategy = "default",
                            bm.format = darwinii_format,
                            calib.lines = cv.k)

user.rf <- list("_allData_allRun"=list(
  mtry=2,
  type = "classification",
  ntre = 250,
  sampsize = round(27 * .95),
  replace = T,
  nodesize=10))
tuned.ann <- bm_Tuning(model = "ANN",
                       tuning.fun = "avNNet",
                       do.formula = T,
                       bm.options = opt.d@options$ANN.binary.nnet.nnet,
                       bm.format = darwinii_format
)
darwinii_opt <- bm_ModelingOptions(
  data.type = "binary",
  models=c("ANN","RF"),
  bm.format = darwinii_format,
  strategy = "user.defined",
  user.base = "bigboss",
  user.val = list(ANN.binary.nnet.nnet = tuned.ann,
                  RF.binary.randomForest.randomForest = user.rf)
)
darwinii_modelo <- BIOMOD_Modeling(
  bm.format = darwinii_format,
  modeling.id = "solo_darwinii",
  models = c("ANN","RF"),
  CV.strategy = "random",
  CV.nb.rep = 5,
  CV.perc = 0.75,
  var.import = 3,
  OPT.strategy = "user.defined",
  OPT.user = darwinii_opt,
  metric.eval = "KAPPA"
)
darwinii_ensemble_models <- BIOMOD_EnsembleModeling(
  bm.mod = darwinii_modelo,
  models.chosen = "all",
  em.by = "all",
  em.algo = c("EMmean", "EMca", "EMcv"),
  metric.select = "KAPPA",
  metric.select.thresh = 0.7,
  metric.eval = "KAPPA",
  var.import = 3
)
-=-=-=-=-=-=-=-=-=-=-=-=-= BIOMOD.ensemble.models.out -=-=-=-=-=-=-=-=-=-=-=-=-=

sp.name : darwinii

expl.var.names : clip_bio3 clip_bio6 clip_bio8 clip_bio9 clip_bio14 clip_bio15 
pedregal_ULTIMO3

models computed: 
darwinii_EMmeanByKAPPA_mergedData_mergedRun_mergedAlgo, darwinii_EMcaByKAPPA_mergedData_mergedRun_mergedAlgo, darwinii_EMcvByKAPPA_mergedData_mergedRun_mergedAlgo

models failed: none

-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
# Proyecciones a futuro######
darwinii_proj <- BIOMOD_Projection(
  bm.mod = darwinii_modelo,
  proj.name = "darwinii",
  new.env = stk_current,
  models.chosen = "all",
  metric.binary = "KAPPA",
  build.clamping.mask = FALSE,
  do.stack = FALSE,
  output.format = ".img")

darwinii_ens_proj <- BIOMOD_EnsembleForecasting(
  bm.em = darwinii_ensemble_models,
  bm.proj = darwinii_proj,
  proj.name = "ENMdarwinii_current",
  models.chosen = 'all',
  metric.binary = 'KAPPA',
  metric.filter = 'all',
  do.stack = FALSE,
  output.format = ".img")

predEMmean <- terra::rast("C:/DATOS/ENM/diplolaemus/darwinii/proj_ENMdarwinii_current/individual_projections/darwinii_EMmeanByKAPPA_mergedData_mergedRun_mergedAlgo.img")

darwinii_proj3 <- BIOMOD_Projection(
  bm.mod = darwinii_modelo,
  proj.name = "future3",
  new.env = stk_future_ssp585_2060,
  models.chosen = "all",
  metric.binary = "KAPPA",
  build.clamping.mask = FALSE,
  do.stack = FALSE,
  output.format = ".img")

darwinii_ens_proj3 <- BIOMOD_EnsembleForecasting(
  bm.em = darwinii_ensemble_models,
  bm.proj = darwinii_proj3,
  proj.name = "ENMdarwinii_future3",
  models.chosen = 'all',
  metric.binary = 'KAPPA',
  metric.filter = 'all',
  do.stack = FALSE,
  output.format = ".img")

predEMmeanFut3 <- terra::rast("C:/DATOS/ENM/diplolaemus/darwinii/proj_ENMdarwinii_future3/individual_projections/darwinii_EMmeanByKAPPA_mergedData_mergedRun_mergedAlgo.img")
``` 
**Environment Information**

> sessionInfo()
R version 4.3.3 (2024-02-29 ucrt)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 19045)

Matrix products: default


locale:
[1] LC_COLLATE=English_United States.utf8  LC_CTYPE=English_United States.utf8    LC_MONETARY=English_United States.utf8
[4] LC_NUMERIC=C                           LC_TIME=English_United States.utf8    

time zone: America/Buenos_Aires
tzcode source: internal

attached base packages:
[1] parallel  splines   stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] caret_6.0-94         sampling_2.10        dplyr_1.1.4          ggplot2_3.5.2        snowfall_1.84-6.3    snow_0.4-4          
 [7] biomod2_4.2-6-1      xgboost_1.7.8.1      randomForest_4.7-1.2 maxnet_0.1.4         earth_5.3.3          plotmo_3.6.4        
[13] plotrix_3.8-4        Formula_1.2-5        gbm_2.2.2            lattice_0.22-5       doParallel_1.0.17    iterators_1.0.14    
[19] ggtext_0.1.2         tidyterra_0.6.1      mgcv_1.9-1           nlme_3.1-166         gam_1.22-5           foreach_1.5.2       
[25] raster_3.6-26        sp_2.1-4             mda_0.5-4            class_7.3-22         rpart_4.1.23         nnet_7.3-19         
[31] terra_1.7-78        

loaded via a namespace (and not attached):
 [1] tidyselect_1.2.1       timeDate_4032.109      reshape_0.8.9          pROC_1.18.5            digest_0.6.35         
 [6] timechange_0.3.0       lifecycle_1.0.4        sf_1.0-16              survival_3.6-4         magrittr_2.0.3        
[11] compiler_4.3.3         rlang_1.1.3            tools_4.3.3            utf8_1.2.4             data.table_1.15.4     
[16] classInt_0.4-10        PresenceAbsence_1.1.11 plyr_1.8.9             xml2_1.3.6             abind_1.4-8           
[21] KernSmooth_2.23-22     withr_3.0.1            purrr_1.0.2            stats4_4.3.3           grid_4.3.3            
[26] fansi_1.0.6            future_1.34.0          e1071_1.7-13           colorspace_2.1-1       globals_0.16.3        
[31] scales_1.3.0           MASS_7.3-60.0.1        cli_3.6.2              generics_0.1.3         future.apply_1.11.2   
[36] rstudioapi_0.16.0      reshape2_1.4.4         DBI_1.2.3              proxy_0.4-27           stringr_1.5.1         
[41] vctrs_0.6.5            hardhat_1.4.0          Matrix_1.6-1.1         jsonlite_1.8.8         listenv_0.9.1         
[46] gower_1.0.1            tidyr_1.3.1            units_0.8-5            recipes_1.1.0          parallelly_1.38.0     
[51] glue_1.7.0             codetools_0.2-19       dismo_1.3-14           lubridate_1.9.3        stringi_1.8.4         
[56] gtable_0.3.5           munsell_0.5.1          tibble_3.2.1           pillar_1.9.0           ipred_0.9-15          
[61] lava_1.8.0             R6_2.5.1               lpSolve_5.6.22         gridtext_0.1.5         Rcpp_1.0.12           
[66] prodlim_2024.06.25     ModelMetrics_1.2.2.2   pkgconfig_2.0.3  
```
Another question is that this warning also appears
Warning message:
executing %dopar% sequentially: no parallel backend registered 

Thanks a lot for your help.
Natalia

**Comment by MayaGueguen** [comment]
Hello there 👋 

There is a `seed.val` parameter in most `BIOMOD_[...]` functions.
If you give it a fixed value, you should get the same results every time 👀 

Maya

**Comment by HeleneBlt** [comment]
HI Natalia, 

The warning : `glm.fit: fitted probabilities numerically 0 or 1 occurred` are not link to the `seed.val`. 

You can ignore safely this message, it just indicates that some values are indistinguishable from 0 and 1 so they will be rounded to 0 or 1. 

Hope it's help 🙂 
Hélène
